functions/scripts/update-algolia-index.ts

```diff
@@ -0,0 +1,78 @@
+import fetch from 'node-fetch';
+import algoliasearch from 'algoliasearch';
+
+if (process.argv.length != 3) throw Error('Term argument not found.');
+
+const term: string = process.argv[2];
+
+if (!/20\d{2}0[1,5,9]/.test(term.trim()))
+  throw Error('Invalid term argument format');
+
+const BASE = 'https://courseup.vikelabs.ca/api';
+
+const client = algoliasearch('ALGOLIA_APP_ID', 'ALGOLIA_ADMIN_KEY');
+const index = client.initIndex('ALGOLIA_INDEX_NAME');
+
+interface Course {
+  subject: string;
+  code: string;
+  pid: string;
+  description: string;
+  title: string;
+}
+
+const getCourses = async (term: string): Promise<Course[]> => {
+  const res = await fetch(`${BASE}/courses/${term}`);
+  const data: Course[] = await res.json();
+  return data;
+};
+
+const getCourse = async (term: string, course: Course): Promise<Course> => {
+  const res = await fetch(
+    `${BASE}/courses/${term}/${course.subject}/${course.code}`
+  );
+  const data: Course = await res.json();
+  return data;
+};
+
+const main = async () => {
+  const courses: Course[] = await getCourses(term);
+
+  console.log(`Found ${courses.length} courses for ${term}`);
+
+  const values = courses.values();
+
+  const vals: Course[] = [];
+
+  const workers = Array(50)
+    .fill(values)
+    .map(async (iterator) => {
+      for (const course of iterator) {
+        const courseDetails = await getCourse(term, course);
+        console.log(course.subject, course.code);
+        vals.push({ ...courseDetails });
+      }
+    });
+
+  await Promise.allSettled(workers);
+
+  const coursesForIndex = vals.map((course) => ({
+    objectID: `${course.subject}${course.code}`,
+    subject: course.subject,
+    code: course.code,
+    pid: course.pid,
+    description: course.description,
+    title: course.title,
+  }));
+
+  index
+    .saveObjects(coursesForIndex)
+    .then(({ objectIDs }) => {
+      console.log(objectIDs);
+    })
+    .catch((err) => {
+      console.log(err);
+    });
```

Comment:
Are you going to update a single index by clearing it and then adding new data or doing this? iirc if you say run this twice in succession you'll have duplicate records in the index. I think the `objectID` should make it so it doesn't but just want to check in.

Reply:
What is the plan for stale courses? ie. courses that don't exist anymore cause this will make it so some will linger.

Reply:
yep you're right! I ran this twice and I don't get duplicate records because of the `objectID` field. What it does is replace _all_ other fields in the object aside from the ID (which in our case will be the subject+code)

but hmmm that's a good point 🤔 wiping it clear and _then_ populating might be good but then it worried me about anything going wrong with the API (e.g. running the script with 202401 currently will return 0 courses to be added since API is not working, so it would simply wipe the index instead of not touching it at all). What do you think?